_upgrading/sopel-8.md

```diff
@@ -0,0 +1,420 @@
+
+Documentation for `@sopel.plugin.command()` suggests alternatives based on the
+use case you need to adapt for Sopel 8.
+
+### Enums everywhere
+
+_(Insert that classic Buzz Lightyear meme here, if you want.)_
+
+Modernizing Sopel to run only on currently-supported Python versions let us do
+quite a bit of work under the hood, but we were also able to start improving how
+some parts of the API work. Using `Enum`s where it makes sense is part of that.
+
+The following parts of Sopel's API are now expressed as some type of `Enum`:
+
+* `sopel.formatting.colors`
+* `sopel.tools.events`
+
+### Identifier casemapping
+
+Any reasonably modern IRC server advertises its method for normalizing nicknames
+to every client that connects. Sopel 8 builds on the `bot.isupport` mechanism
+added in Sopel 7 and makes use of the advertised `CASEMAPPING` value when
+comparing and normalizing nicknames itself.
+
+To support handling this under the hood, the bot's `config.core.nick` setting is
+now stored and returned as a normal `str`. It and any other strings that you'd
+like to treat as nicknames should be passed through the `bot.make_identifier()`
+method to get the same `Identifier` type you're used to from Sopel 7, configured
+to use the IRC server's advertised casemapping method.
+
+### Time-handling changes
+
+`trigger.time` is now an "aware" `datetime` object, meaning it has a UTC offset
+associated with it. Comparison or arithmetic operations between "aware" and
+"naive" `datetime` objects are not allowed; code that manipulates `trigger.time`
+values will need to be updated for Sopel 8.
+
+The fallback format string used by `sopel.tools.time.format_time()` if no other
+source provides one changed from outputting the timezone name (`UTC`) to the UTC
+offset (`+0000`).
+
+`format_time()`'s handling of "aware" and "naive" `datetime`s was also improved,
+but those changes should be transparent to both users and plugin developers.
+
+`sopel.tools.time.validate_timezone()` now also raises a `ValueError` even if
+the timezone to be validated is `None`. This was previously a special case that
+_returned_ `None`.
+
+### Database changes
+
+`bot.db.get_nick_id()` no longer creates a new ID by default if its `create`
+parameter is unspecified.
+
+### IRC event changes
+
+The event names `RPL_INVITELIST` and `RPL_ENDOFINVITELIST` have been updated per
+clarifications from the living-specification project at modern.ircdocs.horse.
+`RPL_INVEXLIST` and `RPL_ENDOFINVEXLIST` were added to the list of named events
+that Sopel knows about.
+
+Plugin code might need to be updated as a result of these changes, but keep in
+mind that [different IRC servers mean different things][invitelist-madness] when
+they send these events. If in doubt, specify a raw numeric value (e.g. `'346'`).
+Writing truly cross-network plugins that react to these events should be
+undertaken _very_ carefully.
+
+  [invitelist-madness]: https://github.com/ircdocs/modern-irc/issues/42
```

Comment:
This addresses multiple related comments, https://github.com/sopel-irc/sopel.chat/pull/46/files#r1548729687 and https://github.com/sopel-irc/sopel.chat/pull/46/files#r1548715208

```suggestion
The event names `RPL_INVITELIST` and `RPL_ENDOFINVITELIST` have been updated per
[clarifications][invitelist-madness] from the "Modern IRC" living-specification project.
`RPL_INVEXLIST` and `RPL_ENDOFINVEXLIST` were added to the list of named events
that Sopel knows about.

Plugin code might need to be updated as a result of these changes, but keep in
mind that [different IRC servers mean different things][invitelist-warning] when
they send these events. If in doubt, specify a raw numeric value (e.g. `'346'`).
Writing truly cross-network plugins that react to these events should be
undertaken _very_ carefully.

  [invitelist-madness]: https://github.com/ircdocs/modern-irc/issues/42
  [invitelist-warning]: https://modern.ircdocs.horse/#invite-list
```

Reply:
Fantastic :smile_cat: